java-runtime/src/test/kotlin/io/spine/validate/NumberRangeSpec.kt

```diff
@@ -0,0 +1,182 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.spine.validate
+
+import com.google.common.truth.Truth.assertThat
+import com.google.protobuf.Message
+import com.google.protobuf.doubleValue
+import io.spine.test.validate.InvalidBound
+import io.spine.test.validate.MaxExclusive
+import io.spine.test.validate.MaxInclusive
+import io.spine.test.validate.MinExclusive
+import io.spine.test.validate.MinInclusive
+import io.spine.type.TypeName
+import io.spine.validate.ValidationOfConstraintTest.Companion.VALIDATION_SHOULD
+import io.spine.validate.given.MessageValidatorTestEnv.VALUE
+import org.junit.jupiter.api.Disabled
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+@DisplayName(VALIDATION_SHOULD + "analyze `(min)` and `(max)` options and")
+internal class NumberRangeSpec : ValidationOfConstraintTest() {
+
+    @Test
+    fun `consider number field is valid if no number options set`() = assertValid {
+        doubleValue { value = 5.0 }
+    }
+
+    @Test
+    fun `find out that number is greater than min inclusive`() =
+        minNumberTest(GREATER_THAN_MIN, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is equal to min inclusive`() =
+        minNumberTest(EQUAL_MIN, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is less than min inclusive`() =
+        minNumberTest(LESS_THAN_MIN, inclusive = true, valid = false)
+
+    @Test
+    fun `find out that number is grated than min exclusive`() =
+        minNumberTest(GREATER_THAN_MIN, inclusive = false, valid = true)
+
+    @Test
+    fun `find out that number is equal to min exclusive`() =
+        minNumberTest(EQUAL_MIN, inclusive = false, valid = false)
+
+    @Test
+    fun `find out that number is less than min exclusive`() =
+        minNumberTest(LESS_THAN_MIN, inclusive = false, valid = false)
+
+    @Test
+    @Disabled("Until 'skipValidation()` is turned off.")
+    fun `provide one valid violation if number is less than min`() {
+        minNumberTest(LESS_THAN_MIN, inclusive = true, valid = false)
+        assertSingleViolation(LESS_MIN_MSG, VALUE)
+    }
+
+    @Test
+    fun `find out that number is greater than max inclusive`() =
+        maxNumberTest(GREATER_THAN_MAX, inclusive = true, valid = false)
+
+    @Test
+    fun `find out that number is equal to max inclusive`() =
+        maxNumberTest(EQUAL_MAX, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is less than max inclusive`() =
+        maxNumberTest(LESS_THAN_MAX, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is less than max non-inclusive`() =
+        maxNumberTest(GREATER_THAN_MAX, inclusive = false, valid = false)
```

Comment:
Can it be more straightforward?

```kotlin
maxNumberTest(LESS_THAN_MAX, inclusive = false, valid = true)
```

Reply:
The first parameter is a value we pass to tests. The `valid` parameter when `true` tells that failure should be asserted.

Here we pass invalid value, and want an error to be raised.